crates/chia-consensus/src/spendbundle_validation.rs

```diff
@@ -46,32 +46,52 @@ pub fn validate_clvm_and_signature(
             .flat_map(move |(condition, items)| {
                 let spend_clone = spend.clone();
                 items.iter().map(move |(pk, msg)| {
-                    (
-                        pk,
-                        make_aggsig_final_message(
-                            condition,
-                            msg.as_slice(),
-                            &spend_clone,
-                            constants,
-                        ),
-                    )
+                    let mut aug_msg = pk.to_bytes().to_vec();
+                    let msg = make_aggsig_final_message(
+                        condition,
+                        msg.as_slice(),
+                        &spend_clone,
+                        constants,
+                    );
+                    aug_msg.extend_from_slice(msg.as_ref());
+                    let aug_hash = hash_to_g2(&aug_msg);
+                    let pairing = aug_hash.pair(pk);
+
+                    (hash_pk_and_msg(&pk.to_bytes(), &msg), pairing)
                 })
             })
     });
-    let unsafe_items = npcresult
-        .agg_sig_unsafe
-        .iter()
-        .map(|(pk, msg)| (pk, msg.as_slice().to_vec()));
+    let unsafes = npcresult.agg_sig_unsafe.clone();
+    let unsafe_items = unsafes.iter().map(|(pk, msg)| {
+        let mut aug_msg = pk.to_bytes().to_vec();
+        aug_msg.extend_from_slice(msg.as_ref());
+        let aug_hash = hash_to_g2(&aug_msg);
+        let pairing = aug_hash.pair(pk);
+        (hash_pk_and_msg(&pk.to_bytes(), msg), pairing)
+    });
     let iter = iter.chain(unsafe_items);
+
     // Verify aggregated signature
-    let (result, added) = cache
-        .lock()
-        .unwrap()
-        .aggregate_verify(iter, &spend_bundle.aggregated_signature);
+    let result = aggregate_verify_gt(
+        &spend_bundle.aggregated_signature,
+        iter.clone().map(|tuple| tuple.1),
```

Comment:
Cloning the iterator here probably means you'll perform the work in the lambda twice, each time you iterate over it. It might be a good reason to `collect()` the `GTElement`s first. Alternatively, build the return vector as part of this lambda function. That would be the most efficient.